drivers/wireless/lpwan/rn2483/rn2483.c

```diff
@@ -0,0 +1,1302 @@
+ * Name: rn2483_read
+ *
+ * Description:
+ *     Character driver interface to RN2483 for receiving data.
+ *
+ ****************************************************************************/
+
+static ssize_t rn2483_read(FAR struct file *filep, FAR char *buffer,
+                           size_t buflen)
+{
+  syslog(LOG_INFO, "Enter read\n");
+
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct rn2483_dev_s *priv = inode->i_private;
+  ssize_t length = 0;
+  int err;
+
+  /* If file position is non-zero, then we're at the end of file. */
+
+  if (filep->f_pos > 0)
+  {
+    return 0;
+  }
+
+  /* Get exclusive access */
+
+  err = nxmutex_lock(&priv->devlock);
+  if (err < 0)
+  {
+    return err;
+  }
+
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+  if (priv->unlinked)
+  {
+    /* Do not allow operations on unlinked drivers. */
+
+    nxmutex_unlock(&priv->devlock);
+    return 0;
+  }
+#endif
+
+  // TODO actually receive
+
+  // For now just get the radio version info
+  char command[] = "sys get ver\r\n";
+  length = file_write(&priv->uart, command, sizeof(command));
+  if (length < 0)
+  {
+    // TODO: log error
+    nxmutex_unlock(&priv->devlock);
+    return length;
+  }
+
+  // Read back the version string
+  length = file_read(&priv->uart, buffer, buflen);
+
+  filep->f_pos += length;
+  nxmutex_unlock(&priv->devlock);
+  return length;
+}
+
+/****************************************************************************
+ * Name: rn2483_write
+ *
+ * Description:
+ *     Write interface for transmitting over the RN2483.
+ ****************************************************************************/
+
+static ssize_t rn2483_write(FAR struct file *filep, FAR const char *buffer, size_t buflen)
+{ // // TODO actually transmit
+
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct rn2483_dev_s *priv = inode->i_private;
+  int err;
+
+  err = nxmutex_lock(&priv->devlock);
+  if (err < 0)
+  {
+    return err;
+  }
+
+  return -ENOSYS;
+}
+
+/****************************************************************************
+ * Name: rn2483_ioctl
+ ****************************************************************************/
+
+static int rn2483_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct rn2483_dev_s *priv = inode->i_private;
+  int err;
+
+  err = nxmutex_lock(&priv->devlock);
+  if (err < 0)
+  {
+    return err;
+  }
+
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+  if (priv->unlinked)
+  {
+    /* Do not allow operations on unlinked drivers. */
+
+    nxmutex_unlock(&priv->devlock);
+    return -ENODEV;
+  }
+#endif
+
+  /* Handle command */
+  switch (cmd)
+  {
+    
+    /*  Set Radio modulation */
+    case WLIOC_SETRADIOMOD:
+    {
+      enum rn2483_mod_e *ptr = (enum rn2483_mod_e *)(arg);
+      if (*ptr != RN2483_MOD_FSK && *ptr != RN2483_MOD_LORA)
+      {
+        return -EINVAL;
```

Comment:
For all your `return -EINVAL` statements, you might want to just set `err = -EINVAL` and then `break;`. If you do an early return here, you need to unlock the mutex first. Instead, by breaking out of the switch statement, you'll arrive at the end of the function and get the mutex unlocked + `return err;` for "free"

Reply:
I just pushed code to add this.